execution-engine/src/wasi/wasmi.rs

```diff
@@ -0,0 +1,1212 @@
+                    not_started,
+                    vec![RuntimeValue::I32(0), RuntimeValue::I32(0)],
+                ),
+                EntrySignature::NoParameters => (not_started, Vec::new()),
+            },
+        };
+
+        not_started.invoke_export(export_name, &program_arguments, self)
+    }
+
+    ////////////////////////////////////////////////////////////////////////////
+    // The WASI host call implementations.
+    ////////////////////////////////////////////////////////////////////////////
+    
+    fn convert_to_errno(input: FileSystemError<()>) -> WasiResult {
+        let errno = match input {
+            Ok(_) => ErrNo::Success,
+            Err(e) => e,
+        };
+        Ok(errno)
+    }
+
+    /// The implementation of the WASI `args_get` function.
+    fn wasi_args_get(&mut self, args: RuntimeArgs) -> WasiResult {
+        TypeCheck::check_args_number(&args, WasiAPIName::ARGS_GET)?;
+        let string_ptr_address = args.nth(0);
+        let buf_address = args.nth(1);
+        Self::convert_to_errno(self
+            .vfs
+            .args_get(&mut self.memory()?, string_ptr_address, buf_address))
+    }
+
+    /// The implementation of the WASI `args_sizes_get` function.
+    fn wasi_args_sizes_get(&mut self, args: RuntimeArgs) -> WasiResult {
+        TypeCheck::check_args_number(&args, WasiAPIName::ARGS_SIZES_GET)?;
+        let arg_count_address: u32 = args.nth(0);
+        let arg_buf_size_address: u32 = args.nth(1);
+        Self::convert_to_errno(self
+            .vfs
+            .args_sizes_get(&mut self.memory()?, arg_count_address, arg_buf_size_address))
+    }
+
+    /// The implementation of the WASI `environ_get` function.
+    fn wasi_environ_get(&mut self, args: RuntimeArgs) -> WasiResult {
+        TypeCheck::check_args_number(&args, WasiAPIName::ENVIRON_GET)?;
+        let string_ptr_address = args.nth(0);
+        let buf_address = args.nth(1);
+        Self::convert_to_errno(self
+            .vfs
+            .environ_get(&mut self.memory()?, string_ptr_address, buf_address))
+    }
+
+    /// The implementation of the WASI `environ_sizes_get` function.
+    fn wasi_environ_sizes_get(&mut self, args: RuntimeArgs) -> WasiResult {
+        TypeCheck::check_args_number(&args, WasiAPIName::ENVIRON_SIZES_GET)?;
+        let environc_address = args.nth::<u32>(0);
+        let environ_buf_size_address = args.nth::<u32>(1);
+        Self::convert_to_errno(self.vfs.environ_sizes_get(
+            &mut self.memory()?,
+            environc_address,
+            environ_buf_size_address,
+        ))
+    }
+
+    /// The implementation of the WASI `clock_res_get` function.  This is not
+    /// supported by Veracruz.  We write `0` as the resolution and return
+    /// `ErrNo::NoSys`.
+    fn wasi_clock_res_get(&mut self, args: RuntimeArgs) -> WasiResult {
+        TypeCheck::check_args_number(&args, WasiAPIName::CLOCK_RES_GET)?;
+        let clock_id = args.nth::<u32>(0);
```

Comment:
The first line of `check_args_number` will check the number of arguments. If the number does not match, it directly return `Err()`.

Reply:
What about the types of the args?

Reply:
I just checked the Wasmi document and source code. The function `nth` will convert the value to the type we want. If fails, it will panic. I think I will update `nth` to `nth_checked`. The latter will convert and return an error if necessary.